lib/spack/spack/parser.py

```diff
@@ -367,10 +367,11 @@ def all_specs(self) -> List["spack.spec.Spec"]:
 class SpecNodeParser:
     """Parse a single spec node from a stream of tokens"""
 
-    __slots__ = "ctx", "has_compiler", "has_version"
+    __slots__ = "ctx", "has_compiler", "has_version", "literal_str"
```

Comment:
Is this new attribute just so that we can print the string as part of the error message? I would think we could print the full token list from `self.ctx` instead for simplicity.

Reply:
The positions in the token are positions in the `literal_str` so that doesn't sound very simple to me. Whitespace could be different if we reconstruct the string from the tokens. Better to refer to the thing that's actually referenced.